src/impression.js

```diff
@@ -53,74 +53,109 @@ export function resetTrackImpressionPromiseForTesting() {
 export function maybeTrackImpression(win) {
   let resolveImpression;
 
-  trackImpressionPromise = new Promise(resolve => {
+  const promise = new Promise(resolve => {
     resolveImpression = resolve;
   });
 
-  const viewer = Services.viewerForDoc(win.document);
-  viewer.isTrustedViewer().then(isTrusted => {
+  trackImpressionPromise = Services.timerFor(win).timeoutPromise(TIMEOUT_VALUE,
+      promise, 'TrackImpressionPromise timeout').catch(error => {
+        dev().warn('IMPRESSION', error);
+      });
+
+  Services.viewerForDoc(win.document).isTrustedViewer().then(isTrusted => {
     // Currently this feature is launched for trusted viewer, but still
     // experiment guarded for all AMP docs.
     if (!isTrusted && !isExperimentOn(win, 'alp')) {
       resolveImpression();
       return;
     }
 
-    trackImpressionFromClickParam(win, viewer, resolveImpression);
+    const replaceUrlPromise = handleReplaceUrl(win);
+    const clickUrlPromise = handleClickUrl(win);
+
+    Promise.all([replaceUrlPromise, clickUrlPromise]).then(() => {
+      resolveImpression();
+    }, () => {});
   });
 }
 
+/**
+ * Signal that impression tracking is not relevant in this environment.
+ */
+export function doNotTrackImpression() {
+  trackImpressionPromise = Promise.resolve();
+}
 
 /**
- * Actually perform the impression request if it has been provided via
- * the click param in the viewer arguments.
+ * Handle the getReplaceUrl and return a promise when url is replaced
+ * Only handles replaceUrl when viewer indicates AMP to do so. Viewer should indicate
+ * by setting the legacy replaceUrl init param and add `replaceUrl` to its capability param.
+ * Future plan is to change the type of legacy init replaceUrl param from url string
+ * to boolean value.
+ * Please NOTE replaceUrl and adLocation will never arrive at same time,
+ * so there is no race condition on the order of handling url replacement.
  * @param {!Window} win
- * @param {!./service/viewer-impl.Viewer} viewer
- * @param {!Function} resolveImpression Call when the impression has been
- *     tracked.
+ * @return {!Promise}
  */
-function trackImpressionFromClickParam(win, viewer, resolveImpression) {
+function handleReplaceUrl(win) {
+  const viewer = Services.viewerForDoc(win.document);
+  if (!viewer.getParam('replaceUrl') || !viewer.hasCapability('replaceUrl')) {
```

Comment:
Did you mean && or ||? Besides, it's not clear why we want to use the init param as well here.

Reply:
I mean `||`.
In https://github.com/ampproject/amphtml/pull/11620#issuecomment-335944455 we decided to use both capability and the init `replaceUrl` param as the indication to ask viewer for async `replaceUrl`.